File: go/day11_part2_2017.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func distance(x, y, z int) int {
	return (abs(x) + abs(y) + abs(z)) / 2
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("File reading error", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	input := scanner.Text()

	directions := strings.Split(input, ",")

	x, y, z := 0, 0, 0
	maxDistance := 0

	for _, dir := range directions {
		switch dir {
		case "n":
			y++
			z--
		case "ne":
			x++
			z--
		case "se":
			x++
			y--
		case "s":
			y--
			z++
		case "sw":
			x--
			z++
		case "nw":
			x--
			y++
		}

		// Calculate the current distance
		curDistance := distance(x, y, z)
		maxDistance = max(maxDistance, curDistance)
	}

	fmt.Println(maxDistance)
}

File: ruby/day11_part2_2017.rb

def abs(x)
  x < 0 ? -x : x
end

def max(a, b)
  a > b ? a : b
end

def distance(x, y, z)
  (abs(x) + abs(y) + abs(z)) / 2
end

file = File.open("input.txt")
input = file.readline.chomp
file.close

directions = input.split(",")

x, y, z = 0, 0, 0
max_distance = 0

directions.each do |dir|
  case dir
  when "n"
    y += 1
    z -= 1
  when "ne"
    x += 1
    z -= 1
  when "se"
    x += 1
    y -= 1
  when "s"
    y -= 1
    z += 1
  when "sw"
    x -= 1
    z += 1
  when "nw"
    x -= 1
    y += 1
  end

  cur_distance = distance(x, y, z)
  max_distance = max(max_distance, cur_distance)
end

puts max_distance
